include/envoy/upstream/cluster_manager.h

```diff
@@ -73,6 +73,15 @@ class ClusterManagerFactory;
 /**
  * Manages connection pools and load balancing for upstream clusters. The cluster manager is
  * persistent and shared among multiple ongoing requests/connections.
+ * Cluster manager is initialed in two phases. In the first phase which begins at the construction
+ * all primary (i.e. not provisioned through xDS) clusters are initialized.
+ * After the first phase the RTDS (if configured) initialization begins. This allows runtime
+ * overrides to be fully configured before the rest of xDS configuration is provisioned.
+ * The second phase of cluster manager initialized begins after RTDS has initialized. In the second
+ * phase all secondary clusters are initialized and then the rest of the configuration provisioned
+ * through xDS.
+ * Please note: this order requires that RTDS is provisioned using a primary cluster. If RTDS is
```

Comment:
No, it is not enforced right now. What would be the best way to do it?

Reply:
There's actually multiple restrictions here:
1. RTDS must be available via a primary cluster.
2. If RTDS happens to be configured with ADS, then ADS must also be available via a primary cluster.
3. Various others, e.g. if a secondary cluster is configured with ADS for its EDS, then ADS must also be available via a primary cluster.

We can enforce these by throwing a config rejection exception on violation of these criteria at construction/config ingest.

Reply:
Yes, the invariants for RTDS config are enforced. The ApiConfigSource must already specified using primary clusters only (checked by the Utility::checkApiConfigSourceSubscriptionBackingCluster). And RTDS provisioned through ADS will fail initialize if ADS is using secondary cluster, since secondary clusters are not present in cluster manager when RTDS is initialized.
I have added server_test tests to check this.